Add registry-watcher image and separate docker-compose services

Nemo157 opened the pull request:
Creates a separate image for the registry-watcher, setup to run the correct command. Reconfigures the docker-compose setup to use the web-server image and this new registry-watcher image as intended to be used in the new production infrastructure, along with a new build-server image that is close to what will be used in the new production infrastructure, replicated twice.

Nemo157 commented:
Probably needs some infra input on the deploy changes (what exactly the tags they use will be). Also I want to investigate the crates.io-index initialization a little more and make sure that's working (and probably have it be local to each container instead of shared as it is now).

syphar commented on dockerfiles/Dockerfile:
```diff
@@ -76,6 +76,66 @@ WORKDIR /srv/docsrs
+
+WORKDIR /srv/docsrs
+# Tini is a small init binary to properly handle signals
+CMD ["/usr/bin/tini", "/usr/local/bin/cratesfyi", "start-registry-watcher"]
```
to replicated the planned production setup, the arguments should include: 
- `repository-stats-updater=enabled` (default: `disabled`) 
- for clarity: `cdn-invalidator=enabled` (default: `enabled`)

syphar replied:
except when infra would like to run these separately, I remember a discussion with pietry where he preferred all these three tasks in one process.

syphar commented on docker-compose.yml:
```diff
@@ -4,33 +4,98 @@ services:
+        depends_on:
+            - db
+        volumes:
+            - "cratesio-index:/opt/docsrs/prefix/crates.io-index"
```
I think this is fine regarding the index, important is only that we test that the registry watcher would clone the index when it doesn't find it.

syphar replied:
Also, assuming that we don't have persistent volumes, currently the index would be re-cloned every time we deploy the container, or restart it. 

So either we should have persistent storage, or perhaps clone the index when building the docker image?

Nemo157 replied:
Yeah, we probably do want to keep it persistent here to avoid recloning all the time, just have it separated from the builders.

syphar commented on docker-compose.yml:
```diff
@@ -4,33 +4,98 @@ services:
+            - s3
+        volumes:
+            - "/var/run/docker.sock:/var/run/docker.sock"
+            - "cratesio-index:/opt/docsrs/prefix/crates.io-index"
```
why does the builder need the index again? What are we using it for? 

`cargo fetch` would use the sparse index protocol, so the only thing I remember is that the builder tries to find the crates.io API URL via the registry, which we could also configure. 

If htere is more we would also have to think about how the builder _updates_ the index, because without the registry watcher there is no update? 

Or am I missing something?

Nemo157 replied:
Just to look up the API url. I think we should change the builder to use the config (which is already there for the search IIRC) instead of reading it from the index. I'll open a separate PR to do that change before this one.

syphar commented on dockerfiles/Dockerfile:
```diff
@@ -76,6 +76,66 @@ WORKDIR /srv/docsrs
+
 ##################
 #  Output stage  #
 ##################
```
after this PR, what are we using the output stage for? 

If for nothing, we could also remove `entrypoint.sh`, which is only used there?

Nemo157 replied:
Nothing that I can think of, I'll clean it up too.

syphar commented:
Thank you for working on this @Nemo157 ! 

Sorry for already adding a review in the draft-stage, but I thought some feedback would already help :)

Nemo157 commented:
I have removed the CI changes pushing the new image to production, since that will require some input from infra and can be done later.

syphar commented on dockerfiles/Dockerfile:
```diff
@@ -65,42 +65,81 @@ RUN apt-get update \
+    && DEBIAN_FRONTEND=noninteractive apt-get upgrade -y \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y \
+        ca-certificates \
+        docker.io \
```
related to the other question above, I wonder if we could limit the CLI stage to just access the same things as the webserver. 

( with the goal in mind that every stage has its own binary with only the dependencies it needs)

syphar commented on docker-compose.yml:
```diff
@@ -1,35 +1,120 @@
+        - s3
+    environment: *environment
+    env_file:
+        - .docker.env
```
why can't we use `.env` here?

Nemo157 replied:
As mentioned below, if `.env` is being used for local settings those are not all appropriate for setting inside docker. It's also loaded by `sqlx` internally through `dotenvy`, so it can't be used to hold settings for docker only and leave local settings to `.envrc` or something (as I do it with `direnv`).

syphar commented on docker-compose.yml:
```diff
@@ -1,35 +1,120 @@
+
+    DOCSRS_PREFIX: /opt/docsrs/prefix
+
+    DOCSRS_DATABASE_URL: postgresql://cratesfyi:password@db
```
not taking this from `.env` would make it harder to mix containerized services with running the services directly (or the CLI). 

Is this intentional?

Nemo157 replied:
Inside containers it needs to be `@db`, outside containers it needs to be `@localhost:15432`, so we can't have a shared `.env` for it anyway. That also somewhat answers the above question, you can't have the containers loading `.env` because that will override the pre-configured variables to use ones for local dev instead of ones for in-container dev.

syphar reviewed (commented):
I wasn't able to fully test yet, but I already have some comments / ideas. 

> Also I want to investigate the crates.io-index initialization a little more and make sure that's working (and probably have it be local to each container instead of shared as it is now).

Were you able to test if the index clone works without the `entrypoint.sh`?

Nemo157 commented on docker-compose.yml:
```diff
@@ -1,35 +1,120 @@
+            - s3
+        volumes:
+            - ".rustwide-docker/cli:/opt/docsrs/rustwide"
+            - "cratesio-index:/opt/docsrs/prefix/crates.io-index"
```
That was how I had it previously, till updating the readme to use `cli` instead of `web` for running `cratesfyi build crate`. I guess I could change that to use a build stage image instead though.

Nemo157 replied:
Updated that, with the `ENTRYPOINT` changes it's now possible to easily use the build-image to run builds (and the `registry-watcher` image to run commands that need the `crates.io-index`).

syphar commented:
short test result: 
- trying to run the registry watcher gives me `/usr/bin/tini: invalid option -- '-'` 
- -

Nemo157 commented:
Fixed that, I was normalizing the `COMMAND`s and thought that the `--` wasn't actually needed with `tini`. I've normalized them the other way and split the `ENTRYPOINT` and `COMMAND` so that it's possible to run other `cratesfyi` commands under any image.

syphar commented:
> Fixed that, I was normalizing the `COMMAND`s and thought that the `--` wasn't actually needed with `tini`. I've normalized them the other way and split the `ENTRYPOINT` and `COMMAND` so that it's possible to run other `cratesfyi` commands under any image.

I'm still seeing the same error when trying to run the registry watcher container?

Nemo157 commented:
> I'm still seeing the same error when trying to run the registry watcher container?

Coming back to this finally, are you using `--build` with the `docker compose up` call? It doesn't have any kind of change detection and requires you to tell it to rebuild. I've added that to the comment about how to run it.

syphar reviewed (changes requested):
tested 
- web container
- cli container 
- registry watcher 
- builder-a


awesome, thank you for keeping to work on it! 

Some small documentation comments, and a missing platform.

Nemo157 commented:
Rebased, and tried to improve the documentation about initial docker compose setup.

Turns out the `start-registry-watcher` command is currently broken though, it tries to use `reqwest::blocking` from within an async task which tokio doesn't like.

syphar commented:
I included your commits in my PR: #2972 

thanks!
